internal/testutil/runmain.go

```diff
@@ -59,46 +63,55 @@ func (r *Runner) Cleanup() {
 	}
 }
 
-// Run runs the built binary with the given environment.
-// After f returns, the running process is shut down.
-func (r *Runner) Run(env map[string]string, f func()) {
+// Stdout returns the stdout from the most recent Run()
+func (r *Runner) Stdout() []byte {
+	return r.stdout.Bytes()
+}
+
+// Stderr returns the stderr from the most recent Run()
+func (r *Runner) Stderr() []byte {
+	return r.stderr.Bytes()
+}
+
+// Run executes runs the built binary until terminated or timeout has
+// been reached, and indicates successful execution on return.
+func (r *Runner) Run(env map[string]string, timeout time.Duration) error {
 	if !r.Built() {
-		r.t.Error("Tried to run when binary not built.")
-		return
+		return fmt.Errorf("tried to run when binary not built")
 	}
+	r.stderr.Reset()
+	r.stdout.Reset()
```

Comment:
I don't particularly like holding so much state on the "Runner" struct.

A separate struct for each run seems to make sense to me, that way you can build once and run several times.

If we only ever need to run once, I'd make a Runner only valid once. This will prevent unexpected concurrency issues.

Reply:
Good point.  I'll cull the state from Runner and return everything directly from Run().